In [1]:
import pandas as pd

In [2]:
import numpy as np

In [3]:
x=np.array([1,2,3,5,6,7,8,9])
np.clip(x,3,8)

array([3, 3, 3, 5, 6, 7, 8, 8])

In [8]:
aTuple = (123, 'xyz', 'zara', 'abc');
aList = list(aTuple)
print(aList[0])

123


In [16]:
#zip 用於將可迭代的對象做為參考，將對應的元素打包成一個個位元組，返回一個列表
# zip[iterable...]
a = [1,2,3]
b = [4,5,6]

zipped = zip(a,b)
print( list(zipped)[0] )

(1, 4)


In [17]:
from itertools import compress
'''
compress 接受兩個參數

- 第一個參數是一個 list, 是你的資料

- 第二個參數我們可以視為 mask (1, 0 表示的 list)
'''
ans = compress('ABCDEF', [1,2,1,2,1,1]) # generator: 直接印看不到結果
ans = list(ans)
ans

['A', 'B', 'C', 'D', 'E', 'F']

In [17]:
scores = [{"name":"A","math":90 , "chinese":50},
         {"name":"B","math":50 , "chinese":80},
         {"name":"C","math":30 , "chinese":93}]
score_df = pd.DataFrame(scores)

In [18]:
score_df

chinese  math name
0       50    90    A
1       80    50    B
2       93    30    C

In [19]:
score_df.set_index("name" , inplace=True)

In [12]:
score_df.loc["A"]

chinese    50
math       90
Name: A, dtype: int64

In [ ]:
score_df.index = {"D" ,"E","F"}
score_df

In [ ]:
score_df.rename(index = {"D":"X" , "E":"Y","F":"Z"})

In [ ]:
score_df['math']>60

In [10]:
score_df.iloc[0]

chinese    50
math       90
Name: F, dtype: int64

In [11]:
score_df.loc["D",['chinese']]

chinese    93
Name: D, dtype: int64

In [12]:
score_df.loc['E'].idxmin()


'math'

In [13]:
score_df.loc['E'].min()

50

In [14]:
score_df["math"].idxmin()

'D'

In [15]:
score_df.idxmin() 

chinese    F
math       D
dtype: object

In [16]:
#我們先讀取成績
math = pd.Series({'小明':90, '小華':80, '小李':70, '小強':55})
chinese = pd.Series({'小明':86, '小華':76, '小李':58, '小強':92})
english = pd.Series({'小明':0, '小華':45, '小李':69, '小強':32})

In [17]:
def rank(sr , subject):
    new_sr = sr.to_frame()
    new_sr = new_sr.rename(columns = {0:subject})
    new_sr[subject+"_rank"] = new_sr[subject].rank(method="min",ascending=False)
    return new_sr

In [18]:
#處理這三個科目
math = rank(math, '數學')
chinese = rank(chinese, '國語')
english = rank(english, '英文')
sum_table = pd.concat([math,chinese,english], axis=1) #合併

In [19]:
sum_table['總分'] = sum_table['數學']+sum_table['國語']+sum_table['英文']
sum_table["總排名"] = sum_table["總分"].rank(method="min",ascending=False)
sum_table = sum_table.sort_values("總排名")
sum_table

數學  數學_rank  國語  國語_rank  英文  英文_rank   總分  總排名
小華  80      2.0  76      3.0  45      2.0  201  1.0
小李  70      3.0  58      4.0  69      1.0  197  2.0
小強  55      4.0  92      1.0  32      3.0  179  3.0
小明  90      1.0  86      2.0   0      4.0  176  4.0

In [20]:
def function(a,b):
    print(a,b)

In [21]:
g = lambda x : x ** 2
print (g(3))

9


In [22]:
g = lambda x,y,z :(x+y)**z
print(g(1,2,3) )

27


In [23]:
a = [1,2,3]

In [24]:
map(lambda x : x*2 , a)
a

[1, 2, 3]

In [25]:
print map(lambda x : x*2, [1,2,3,4,5]) 

SyntaxError: invalid syntax (<ipython-input-25-6ad844386432>, line 1)

In [26]:
def multiple2(x):
    return x*2

In [27]:
list1 = [1,3,5,7,9]
print( list(map(multiple2,list1)) )

[2, 6, 10, 14, 18]


In [28]:
map()函数不改变原有的 list，而是返回一个新的 list。

SyntaxError: invalid syntax (<ipython-input-28-cd20496a47dd>, line 1)

In [29]:
import random

In [30]:
print(random.choice("tmrrw"))

r


In [31]:
for i in range(5):
    print( random.randrange(0,100,2) ) 

32
38
16
96
36


In [32]:
a = [1,3,5,7,9]
random.shuffle(a)
print(a)

[9, 1, 5, 7, 3]


In [33]:
for i in range(5):
    print (random.sample('abcde',5) )

['a', 'c', 'b', 'd', 'e']
['d', 'a', 'e', 'c', 'b']
['e', 'd', 'b', 'c', 'a']
['e', 'd', 'a', 'c', 'b']
['a', 'b', 'e', 'c', 'd']


In [34]:
import numpy as np

In [35]:
# 在默认情况下，linspace函数可以生成元素为50的等间隔数列
x1 = np.linspace(1,10)

In [36]:
print(x1)

[ 1.          1.18367347  1.36734694  1.55102041  1.73469388  1.91836735
  2.10204082  2.28571429  2.46938776  2.65306122  2.83673469  3.02040816
  3.20408163  3.3877551   3.57142857  3.75510204  3.93877551  4.12244898
  4.30612245  4.48979592  4.67346939  4.85714286  5.04081633  5.2244898
  5.40816327  5.59183673  5.7755102   5.95918367  6.14285714  6.32653061
  6.51020408  6.69387755  6.87755102  7.06122449  7.24489796  7.42857143
  7.6122449   7.79591837  7.97959184  8.16326531  8.34693878  8.53061224
  8.71428571  8.89795918  9.08163265  9.26530612  9.44897959  9.63265306
  9.81632653 10.        ]


In [37]:
print("length if x1 is %d" %len(x1))

length if x1 is 50


In [44]:
A = [2,1,2,3,4,5,6,7,7]
uniA = np.unique(A)
print(uniA)

[1 2 3 4 5 6 7]


In [45]:
# return_index=True表示返回新列表元素在旧列表中的位置，并以列表形式储存在s中。
uniA2, uniA3 = np.unique(A,return_index=True)
print(uniA2)
print(uniA3)

[1 2 3 4 5 6 7]
[1 0 3 4 5 6 7]


In [46]:
# 返回旧列表元素在新列表中的位置，并以列表形式储存在p中
print(uniA4)
print(uniA5)
print(uniA6)

[1 2 3 4 5 6 7]
[1 0 3 4 5 6 7]
[1 0 1 2 3 4 5 6 6]


#### predict_proba返回的是一个n行k列的数组
#### 第i行第j列上的数值是模型预测第i个预测样本的标签为j的概率
#### 所以每一行的和应该等于1.

In [1]:
from sklearn.linear_model import LogisticRegression
import numpy as np

In [2]:
x_train = np.array([[1,2,3],[1,3,4],[2,1,2],[4,5,6],[3,5,3],[1,7,2]])
y_train = np.array([0, 0, 0, 1, 1, 1])
x_test = np.array([[2,2,2],[3,2,6],[1,7,4]])

In [3]:
clf = LogisticRegression()
clf.fit(x_train , y_train)

C:\Users\lincecum\AppData\Roaming\Python\Python36\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=None, solver='warn', tol=0.0001, verbose=0,
                   warm_start=False)

In [4]:
# 返回预测标签
clf.predict(x_test)

array([1, 0, 1])

In [5]:
# 返回预测属于某标签的概率
clf.predict_proba(x_test)

array([[0.43348191, 0.56651809],
       [0.84401838, 0.15598162],
       [0.13147498, 0.86852502]])

### 预测[2,2,2]的标签是0的概率为0.43348191，1的概率为0.56651809 <br/>预测[3,2,6]的标签是0的概率为0.84401838，1的概率为0.15598162 <br/>預測[1,7,4]的标签是0的概率为0.13147498，1的概率为0.86852502

In [ ]:
probas是一个k列的array

probas[:, j]这一列是每个test样本为j的概率。

probas[i, j]是预测第i个样本的标签为j的概率。

In [18]:
from sklearn.datasets import load_iris
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import Normalizer
from sklearn.preprocessing import Binarizer
from sklearn.preprocessing import OneHotEncoder

In [7]:
# 導入數據集
iris = load_iris()


In [8]:
# 特徵矩陣
iris.data

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
       [4.9, 3

In [9]:
# 目標向量
iris.target

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [11]:
StandardScaler().fit_transform(iris.data)

array([[-9.00681170e-01,  1.01900435e+00, -1.34022653e+00,
        -1.31544430e+00],
       [-1.14301691e+00, -1.31979479e-01, -1.34022653e+00,
        -1.31544430e+00],
       [-1.38535265e+00,  3.28414053e-01, -1.39706395e+00,
        -1.31544430e+00],
       [-1.50652052e+00,  9.82172869e-02, -1.28338910e+00,
        -1.31544430e+00],
       [-1.02184904e+00,  1.24920112e+00, -1.34022653e+00,
        -1.31544430e+00],
       [-5.37177559e-01,  1.93979142e+00, -1.16971425e+00,
        -1.05217993e+00],
       [-1.50652052e+00,  7.88807586e-01, -1.34022653e+00,
        -1.18381211e+00],
       [-1.02184904e+00,  7.88807586e-01, -1.28338910e+00,
        -1.31544430e+00],
       [-1.74885626e+00, -3.62176246e-01, -1.34022653e+00,
        -1.31544430e+00],
       [-1.14301691e+00,  9.82172869e-02, -1.28338910e+00,
        -1.44707648e+00],
       [-5.37177559e-01,  1.47939788e+00, -1.28338910e+00,
        -1.31544430e+00],
       [-1.26418478e+00,  7.88807586e-01, -1.22655167e+00,
      

In [13]:
# 区间缩放，返回值为缩放到[0, 1]区间的数据
MinMaxScaler().fit_transform(iris.data)   

array([[0.22222222, 0.625     , 0.06779661, 0.04166667],
       [0.16666667, 0.41666667, 0.06779661, 0.04166667],
       [0.11111111, 0.5       , 0.05084746, 0.04166667],
       [0.08333333, 0.45833333, 0.08474576, 0.04166667],
       [0.19444444, 0.66666667, 0.06779661, 0.04166667],
       [0.30555556, 0.79166667, 0.11864407, 0.125     ],
       [0.08333333, 0.58333333, 0.06779661, 0.08333333],
       [0.19444444, 0.58333333, 0.08474576, 0.04166667],
       [0.02777778, 0.375     , 0.06779661, 0.04166667],
       [0.16666667, 0.45833333, 0.08474576, 0.        ],
       [0.30555556, 0.70833333, 0.08474576, 0.04166667],
       [0.13888889, 0.58333333, 0.10169492, 0.04166667],
       [0.13888889, 0.41666667, 0.06779661, 0.        ],
       [0.        , 0.41666667, 0.01694915, 0.        ],
       [0.41666667, 0.83333333, 0.03389831, 0.04166667],
       [0.38888889, 1.        , 0.08474576, 0.125     ],
       [0.30555556, 0.79166667, 0.05084746, 0.125     ],
       [0.22222222, 0.625     ,

In [15]:
Normalizer().fit_transform(iris.data)

array([[0.80377277, 0.55160877, 0.22064351, 0.0315205 ],
       [0.82813287, 0.50702013, 0.23660939, 0.03380134],
       [0.80533308, 0.54831188, 0.2227517 , 0.03426949],
       [0.80003025, 0.53915082, 0.26087943, 0.03478392],
       [0.790965  , 0.5694948 , 0.2214702 , 0.0316386 ],
       [0.78417499, 0.5663486 , 0.2468699 , 0.05808704],
       [0.78010936, 0.57660257, 0.23742459, 0.0508767 ],
       [0.80218492, 0.54548574, 0.24065548, 0.0320874 ],
       [0.80642366, 0.5315065 , 0.25658935, 0.03665562],
       [0.81803119, 0.51752994, 0.25041771, 0.01669451],
       [0.80373519, 0.55070744, 0.22325977, 0.02976797],
       [0.786991  , 0.55745196, 0.26233033, 0.03279129],
       [0.82307218, 0.51442011, 0.24006272, 0.01714734],
       [0.8025126 , 0.55989251, 0.20529392, 0.01866308],
       [0.81120865, 0.55945424, 0.16783627, 0.02797271],
       [0.77381111, 0.59732787, 0.2036345 , 0.05430253],
       [0.79428944, 0.57365349, 0.19121783, 0.05883625],
       [0.80327412, 0.55126656,

In [17]:
#二值化，阈值设置为3，返回值为二值化后的数据
Binarizer(threshold = 3).fit_transform(iris.data)

array([[1., 1., 0., 0.],
       [1., 0., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 0., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 0., 0., 0.],
       [1., 0., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 0., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 0., 0., 0.],
       [1., 1., 0., 0.],


In [19]:
#哑编码，对IRIS数据集的目标值，返回值为哑编码后的数据
OneHotEncoder().fit_transform(iris.target.reshape(-1,1))

C:\Users\lincecum\AppData\Roaming\Python\Python36\site-packages\sklearn\preprocessing\_encoders.py:415: FutureWarning: The handling of integer data will change in version 0.22. Currently, the categories are determined based on the range [0, max(values)], while in the future they will be determined based on the unique values.
If you want the future behaviour and silence this warning, you can specify "categories='auto'".
In case you used a LabelEncoder before this OneHotEncoder to convert the categories to integers, then you can now use the OneHotEncoder directly.
  warnings.warn(msg, FutureWarning)


<150x3 sparse matrix of type '<class 'numpy.float64'>'
	with 150 stored elements in Compressed Sparse Row format>

In [20]:
from sklearn.feature_selection import VarianceThreshold

c:\program files\python36\lib\importlib\_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192 from C header, got 216 from PyObject
  return f(*args, **kwds)


In [21]:
#方差选择法，返回值为特征选择后的数据
#参数threshold为方差的阈值
VarianceThreshold(threshold= 3).fit_transform(iris.data)

array([[1.4],
       [1.4],
       [1.3],
       [1.5],
       [1.4],
       [1.7],
       [1.4],
       [1.5],
       [1.4],
       [1.5],
       [1.5],
       [1.6],
       [1.4],
       [1.1],
       [1.2],
       [1.5],
       [1.3],
       [1.4],
       [1.7],
       [1.5],
       [1.7],
       [1.5],
       [1. ],
       [1.7],
       [1.9],
       [1.6],
       [1.6],
       [1.5],
       [1.4],
       [1.6],
       [1.6],
       [1.5],
       [1.5],
       [1.4],
       [1.5],
       [1.2],
       [1.3],
       [1.4],
       [1.3],
       [1.5],
       [1.3],
       [1.3],
       [1.3],
       [1.6],
       [1.9],
       [1.4],
       [1.6],
       [1.4],
       [1.5],
       [1.4],
       [4.7],
       [4.5],
       [4.9],
       [4. ],
       [4.6],
       [4.5],
       [4.7],
       [3.3],
       [4.6],
       [3.9],
       [3.5],
       [4.2],
       [4. ],
       [4.7],
       [3.6],
       [4.4],
       [4.5],
       [4.1],
       [4.5],
       [3.9],
       [4.8],
      

In [22]:
from sklearn.feature_selection import SelectKBest
from scipy.stats import pearsonr

In [ ]:
#选择K个最好的特征，返回选择特征后的数据
#第一个参数为计算评估特征是否好的函数，该函数输入特征矩阵和目标向量，输出二元组（评分，P值）的数组，数组第i项为第i个特征的评分和P值。在此定义为计算相关系数
#参数k为选择的特征个数
SelectKBest(lambda X, Y: array(map(lambda x:pearsonr(x, Y), X.T)).T, k=2).fit_transform(iris.data, iris.target)


In [25]:
from sklearn.feature_selection import chi2

In [26]:
#选择K个最好的特征，返回选择特征后的数据
SelectKBest(chi2, k=2).fit_transform(iris.data, iris.target)

array([[1.4, 0.2],
       [1.4, 0.2],
       [1.3, 0.2],
       [1.5, 0.2],
       [1.4, 0.2],
       [1.7, 0.4],
       [1.4, 0.3],
       [1.5, 0.2],
       [1.4, 0.2],
       [1.5, 0.1],
       [1.5, 0.2],
       [1.6, 0.2],
       [1.4, 0.1],
       [1.1, 0.1],
       [1.2, 0.2],
       [1.5, 0.4],
       [1.3, 0.4],
       [1.4, 0.3],
       [1.7, 0.3],
       [1.5, 0.3],
       [1.7, 0.2],
       [1.5, 0.4],
       [1. , 0.2],
       [1.7, 0.5],
       [1.9, 0.2],
       [1.6, 0.2],
       [1.6, 0.4],
       [1.5, 0.2],
       [1.4, 0.2],
       [1.6, 0.2],
       [1.6, 0.2],
       [1.5, 0.4],
       [1.5, 0.1],
       [1.4, 0.2],
       [1.5, 0.2],
       [1.2, 0.2],
       [1.3, 0.2],
       [1.4, 0.1],
       [1.3, 0.2],
       [1.5, 0.2],
       [1.3, 0.3],
       [1.3, 0.3],
       [1.3, 0.2],
       [1.6, 0.6],
       [1.9, 0.4],
       [1.4, 0.3],
       [1.6, 0.2],
       [1.4, 0.2],
       [1.5, 0.2],
       [1.4, 0.2],
       [4.7, 1.4],
       [4.5, 1.5],
       [4.9,

In [27]:
from sklearn.feature_selection import SelectFromModel
from sklearn.ensemble import GradientBoostingClassifier

c:\program files\python36\lib\importlib\_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192 from C header, got 216 from PyObject
  return f(*args, **kwds)


In [28]:
#GBDT作为基模型的特征选择
SelectFromModel(GradientBoostingClassifier()).fit_transform(iris.data , iris.target)

array([[0.2],
       [0.2],
       [0.2],
       [0.2],
       [0.2],
       [0.4],
       [0.3],
       [0.2],
       [0.2],
       [0.1],
       [0.2],
       [0.2],
       [0.1],
       [0.1],
       [0.2],
       [0.4],
       [0.4],
       [0.3],
       [0.3],
       [0.3],
       [0.2],
       [0.4],
       [0.2],
       [0.5],
       [0.2],
       [0.2],
       [0.4],
       [0.2],
       [0.2],
       [0.2],
       [0.2],
       [0.4],
       [0.1],
       [0.2],
       [0.2],
       [0.2],
       [0.2],
       [0.1],
       [0.2],
       [0.2],
       [0.3],
       [0.3],
       [0.2],
       [0.6],
       [0.4],
       [0.3],
       [0.2],
       [0.2],
       [0.2],
       [0.2],
       [1.4],
       [1.5],
       [1.5],
       [1.3],
       [1.5],
       [1.3],
       [1.6],
       [1. ],
       [1.3],
       [1.4],
       [1. ],
       [1.5],
       [1. ],
       [1.4],
       [1.3],
       [1.4],
       [1.5],
       [1. ],
       [1.5],
       [1.1],
       [1.8],
      

In [29]:
from sklearn.decomposition import PCA

In [30]:
#主成分分析法，返回降维后的数据
#参数n_components为主成分数目
PCA(n_components=2).fit_transform(iris.data)

array([[-2.68412563,  0.31939725],
       [-2.71414169, -0.17700123],
       [-2.88899057, -0.14494943],
       [-2.74534286, -0.31829898],
       [-2.72871654,  0.32675451],
       [-2.28085963,  0.74133045],
       [-2.82053775, -0.08946138],
       [-2.62614497,  0.16338496],
       [-2.88638273, -0.57831175],
       [-2.6727558 , -0.11377425],
       [-2.50694709,  0.6450689 ],
       [-2.61275523,  0.01472994],
       [-2.78610927, -0.235112  ],
       [-3.22380374, -0.51139459],
       [-2.64475039,  1.17876464],
       [-2.38603903,  1.33806233],
       [-2.62352788,  0.81067951],
       [-2.64829671,  0.31184914],
       [-2.19982032,  0.87283904],
       [-2.5879864 ,  0.51356031],
       [-2.31025622,  0.39134594],
       [-2.54370523,  0.43299606],
       [-3.21593942,  0.13346807],
       [-2.30273318,  0.09870885],
       [-2.35575405, -0.03728186],
       [-2.50666891, -0.14601688],
       [-2.46882007,  0.13095149],
       [-2.56231991,  0.36771886],
       [-2.63953472,

In [8]:
country = ['A' , 'B' , 'C' , 'D']
age = [12,34,54,29]
salary = [30000,40000,60000,20000]
dic = {'Country':country , 'Age':age , 'Salary':salary}
data = pd.DataFrame(dic)
data

Age Country  Salary
0   12       A   30000
1   34       B   40000
2   54       C   60000
3   29       D   20000

In [5]:
# label encoding
from sklearn.preprocessing import LabelEncoder
labelencoder = LabelEncoder()
data_le = pd.DataFrame(dic)
data_le['Country'] = labelencoder.fit_transform(data_le['Country'])
data_le

Age  Country  Salary
0   12        0   30000
1   34        1   40000
2   54        2   60000
3   29        3   20000

In [6]:
from sklearn.preprocessing import OneHotEncoder

In [7]:
data_dum = pd.get_dummies(data)
pd.DataFrame(data_dum)

Age  Salary  Country_A  Country_B  Country_C  Country_D
0   12   30000          1          0          0          0
1   34   40000          0          1          0          0
2   54   60000          0          0          1          0
3   29   20000          0          0          0          1

In [1]:
import pandas as pd
df = pd.DataFrame({'Q':['LI','ZHANG','ZHANG','LI','WANG'], 'A' : [1,1,1,2,2], 'B' : [1,-1,0,1,2], 'C' : [3,4,5,6,7]})

In [2]:
df

A  B  C      Q
0  1  1  3     LI
1  1 -1  4  ZHANG
2  1  0  5  ZHANG
3  2  1  6     LI
4  2  2  7   WANG

In [3]:
df.groupby("Q").apply(lambda x:print(x))

   A  B  C   Q
0  1  1  3  LI
3  2  1  6  LI
   A  B  C   Q
0  1  1  3  LI
3  2  1  6  LI
   A  B  C     Q
4  2  2  7  WANG
   A  B  C      Q
1  1 -1  4  ZHANG
2  1  0  5  ZHANG


Empty DataFrame
Columns: []
Index: []

In [4]:
df.groupby("Q").agg(lambda x:print(x))

0    1
3    2
Name: A, dtype: int64
4    2
Name: A, dtype: int64
1    1
2    1
Name: A, dtype: int64
0    1
3    1
Name: B, dtype: int64
4    2
Name: B, dtype: int64
1   -1
2    0
Name: B, dtype: int64
0    3
3    6
Name: C, dtype: int64
4    7
Name: C, dtype: int64
1    4
2    5
Name: C, dtype: int64


A     B     C
Q                      
LI     None  None  None
WANG   None  None  None
ZHANG  None  None  None

In [5]:
df.groupby("Q").agg(["mean" ,"std","count","max"])

A                        B                        C                  \
      mean       std count max mean       std count max mean       std count   
Q                                                                              
LI     1.5  0.707107     2   2  1.0  0.000000     2   1  4.5  2.121320     2   
WANG   2.0       NaN     1   2  2.0       NaN     1   2  7.0       NaN     1   
ZHANG  1.0  0.000000     2   1 -0.5  0.707107     2   0  4.5  0.707107     2   

           
      max  
Q          
LI      6  
WANG    7  
ZHANG   5